src/VisualStudio/LiveShare/Impl/RenameHandler.cs

```diff
@@ -0,0 +1,81 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.ComponentModel.Composition;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Editor;
+using Microsoft.CodeAnalysis.Editor.Shared.Utilities;
+using Microsoft.CodeAnalysis.LanguageServer;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+using Microsoft.VisualStudio.LiveShare.LanguageServices;
+
+namespace Microsoft.VisualStudio.LanguageServices.LiveShare
+{
+    [ExportLspRequestHandler(LiveShareConstants.RoslynContractName, Methods.TextDocumentRenameName)]
+    internal class RenameHandler : ILspRequestHandler<RenameParams, WorkspaceEdit, Solution>
+    {
+        private readonly IThreadingContext _threadingContext;
+
+        [ImportingConstructor]
+        public RenameHandler(IThreadingContext threadingContext)
+        {
+            _threadingContext = threadingContext;
+        }
+
+        public async Task<WorkspaceEdit> HandleAsync(RenameParams request, RequestContext<Solution> requestContext, CancellationToken cancellationToken)
+        {
+            var solution = requestContext.Context;
+            WorkspaceEdit workspaceEdit = null;
+            var document = solution.GetDocumentFromURI(request.TextDocument.Uri);
+            if (document != null)
+            {
+                var renameService = document.Project.LanguageServices.GetService<IEditorInlineRenameService>();
+                var position = await document.GetPositionFromLinePositionAsync(ProtocolConversions.PositionToLinePosition(request.Position), cancellationToken).ConfigureAwait(false);
+
+                // We need to be on the UI thread to call GetRenameInfo which computes the rename locations.
+                // This is because Roslyn reads the readonly regions of the buffer to compute the locations in the document.
+                // This is typically quick. It's marked configureawait(false) so that the bulk of the rename operation can happen
+                // in background threads.
+                await _threadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
```

Comment:
let's have work item to figure out how to move this down to LSP layer which doesn't require UI.

Reply:
https://github.com/dotnet/roslyn/projects/45#card-20619668